### DISCUSSION TITLE
Incorrect extraction of the number of values.

### DISCUSSION THREAD

**Discussion by zusage** [BurntSushi/ripgrep#2522] [Q&A] [answered]
When getting values, the following occurs.
Perhaps I'm using the syntax incorrectly.

$ echo 'foo(var1, var2, var3, var4, var5)' | rg '^foo\((([^,]+,?){5})\)' -r '$1'  
var1, var2, var3, var4, var5   <--- Correct, since the number of required matches is {5}
Expected result: var1, var2, var3, var4, var5

$ echo 'foo(var1, var2, var3, var4, var5)' | rg '^foo\((([^,]+,?){4})\)' -r '$1'
Nothing   <--- Correct, since the number of required matches is {4}
Expected result: Nothing

$ echo 'foo(var1, var2, var3, var4, var5)' | rg '^foo\((([^,]+,?){6})\)' -r '$1'
var1, var2, var3, var4, var5   <--- Incorrect, since the number of required matches is {6}
Expected result: Nothing

**Comment by BurntSushi** [comment, marked as answer]
It's correct. I believe you're misunderstand regex semantics here. A simpler example:

```
$ echo 'aaaaa' | rg '(a+){2}'
aaaaa
$ echo 'aaaaa' | rg '(a+){3}'
aaaaa
$ echo 'aaaaa' | rg '(a+){4}'
aaaaa
$ echo 'aaaaa' | rg '(a+){5}'
aaaaa
$ echo 'aaaaa' | rg '(a+){6}'
$
```

That is, ripgrep will prioritize a match over greediness. So that means a `(re)+` can result in matching fewer repetitions of `re` if it means an overall match could be found. In your regex, `^foo((([^,]+,?){9}))`, it might be helpful to focus on the thing being repeated: `[^,]+,?`. Notice that that regex can match *any* single letter except for `,` since the `,` at the end is made optional by `?`.

I believe what you want is to make sure you match a terminator:

```
$ echo 'foo(var1, var2, var3, var4, var5)' | rg '^foo((([^,]+(,|\))){5}))' -r '$1'
(var1, var2, var3, var4, var5)
$ echo 'foo(var1, var2, var3, var4, var5)' | rg '^foo((([^,]+(,|\))){6}))' -r '$1'
```

**Comment by zusage** [comment]
Could you please check the results? Perhaps I missed something important.
$ echo 'foo(var1, var2, var3, var4, var5)' | rg '^foo((([^,]+(,|\))){5}))' -r '$1'
Expected result: var1, var2, var3, var4, var5

$ echo 'foo(var1, var2, var3, var4, var5)' | rg '^foo((([^,]+(,|\))){4}))' -r '$1'
Expected result: Nothing

$ echo 'foo(var1, var2, var3, var4, var5)' | rg '^foo((([^,]+(,|\))){6}))' -r '$1'
Expected result: Nothing

**Reply by BurntSushi** [reply]
`,|` is effectively equivalent to `,?`. So you have the same problem as your original regex.

**Reply by BurntSushi** [reply]
Okay special attention to the fact that `\(` matches a literal `(`. That is, it escapes the parenthesis. I assume you meant to escape some of the parentheses, but I don't see any escapes in your regex.
